---
title: 'Learn: Package `TCGAWorkflow`'
author: 'Leo Williams'
# output: html_notebook
---

# ABOUT

This notebook is for learning the R Bioconductor package `TCGAWorkflow`. It follows the vigntte [TCGA Workflow: Analyze cancer genomics and epigenomics data using Bioconductor packages](https://www.bioconductor.org/packages/release/workflows/vignettes/TCGAWorkflow/inst/doc/TCGAWorkflow.html) from the package [webpage](https://www.bioconductor.org/packages/release/workflows/html/TCGAWorkflow.html).

Data used is from the package `TCGAWorkflowData`.

Other resources:

-   Workshop -[TCGA data analysis](http://rstudio-pubs-static.s3.amazonaws.com/523337_fb0f67ec236b413486f9852a8263809e.html#43_mutation)

-   

# SETUP

Install packages if needed:

```{r install-packages}
if (FALSE) {
	BiocManager::install(c(
		"TCGAWorkflow"
	))
}
```

Load packages and set options:

```{r setup, message=FALSE}
library(TCGAbiolinks)
library(TCGAWorkflowData)
library(DT)
library(tidyverse)
```

# INTRODUCTION

# METHODS

## Access to the data

### Downloading data from TCGA data portal

`TCGAbiolinks` has 3 main functions:

-   `GDCquery()`
-   `GDCdownload()`
-   `GDCprepare()`

**`GDCquery()`** uses GDC API to search the data for a given project and data category and filters results by features requested. Returns a summary table. Most important args are `peoject`, `data.category`, `data.type`, `workflow.type`, `legacy`, `file.type`, and `platform`. A complete list of possible arguments can be found in the [`TCGAbiolinks` vignette](https://www.bioconductor.org/packages/3.3/bioc/vignettes/TCGAbiolinks/inst/doc/tcgaBiolinks.html#harmonized-data-1).

**`GDCdownload()`** is used to download the data, or the `gdc-client` can be used. The downloaded data will be saved in a dir with the project name and a subfolder with the `data.cetegory`.

**`GDCprepare()`** transforms downloaded data into a `summarizedExperiment` object or a data frame. If `SummarizedExperiment = TRUE`, TCGAbiolinks will add to the object sub-type information (defined by TCGA) and clinical information.

```{r query-and-download-data}
# Obs: The data in the legacy database has been aligned to hg19
query.met.gbm <- GDCquery(project = "TCGA-GBM", 
                          legacy = TRUE,
                          data.category = "DNA methylation",
                          platform = "Illumina Human Methylation 450", 
                          barcode = c("TCGA-76-4926-01B-01D-1481-05", "TCGA-28-5211-01C-11D-1844-05"))
GDCdownload(query.met.gbm)

met.gbm.450 <- GDCprepare(query = query.met.gbm,
                          save = TRUE, 
                          save.filename = "gbmDNAmet450k.rda",
                          summarizedExperiment = TRUE)

query.met.lgg <- GDCquery(project = "TCGA-LGG", 
                          legacy = TRUE,
                          data.category = "DNA methylation",
                          platform = "Illumina Human Methylation 450",
                          barcode = c("TCGA-HT-7879-01A-11D-2399-05", "TCGA-HT-8113-01A-11D-2399-05"))
GDCdownload(query.met.lgg)
met.lgg.450 <- GDCprepare(query = query.met.lgg,
                          save = TRUE, 
                          save.filename = "lggDNAmet450k.rda",
                          summarizedExperiment = TRUE)
met.gbm.lgg <- SummarizedExperiment::cbind(met.lgg.450, met.gbm.450)


query.exp.lgg <- GDCquery(project = "TCGA-LGG", 
                          legacy = TRUE,
                          data.category = "Gene expression",
                          data.type = "Gene expression quantification",
                          platform = "Illumina HiSeq", 
                          file.type = "results",
                          sample.type = "Primary solid Tumor")
GDCdownload(query.exp.lgg)
exp.lgg <- GDCprepare(query = query.exp.lgg, save = TRUE, save.filename = "lggExp.rda")

query.exp.gbm <- GDCquery(project = "TCGA-GBM", 
                          legacy = TRUE,
                          data.category = "Gene expression",
                          data.type = "Gene expression quantification",
                          platform = "Illumina HiSeq", 
                          file.type = "results",
                          sample.type = "Primary solid Tumor")
GDCdownload(query.exp.gbm)
exp.gbm <- GDCprepare(query = query.exp.gbm, save = TRUE, save.filename = "gbmExp.rda")
exp.gbm.lgg <- SummarizedExperiment::cbind(exp.lgg, exp.gbm)
```

# TEST

Query GDC for annotated Varscan VCFs for CRC and UCEC tumors:

```{r query-CRC-UCEC}
# FIXME: does not return case ID
query_snv_crc_ucec = GDCquery(project = c("TCGA-COAD", 
										  "TCGA-READ", 
										  "TCGA-UCEC"), 
							  data.category = "Simple Nucleotide Variation", 
							  data.type = "Annotated Somatic Mutation", 
							  workflow.type = "VarScan2 Annotation", 
							  data.format = "VCF")

```

# TO-DO

-   [x] ~~can you only search one project at a time? -- no, you can search all, but every project included must *have* the type of data being requested (e.g.: "Error in checkDataCategoriesInput(project, data.category, legacy) : Please set a valid data.category argument from the column data_category above. We could not validade the data.category for project GENIE-MSK")~~
